### ISSUE TITLE
TestInOutVariables.py: '(lldb) po x' crashing on first breakpoint

### ISSUE THREAD

**Issue by tfiala** [tfiala/swift-lldb#1]
Here's how to reproduce:
```
$ cd lldb/packages/Python/lldbsuite/test/lang/swift/variables/inout
$ make
$ ls
a.out  core  main.swift  Makefile  TestInOutVariables.py
$ ulimit -c unlimited
$ ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb a.out
(lldb) target create "a.out"
Current executable set to 'a.out' (x86_64).
(lldb) breakpoint set -p 'Set breakpoint here for Class access'
Breakpoint 1: where = a.out`a.foo (inout a.Class) -> () + 189 at main.swift:43, address = 0x000000000040140d
(lldb) run
Process 15955 launched: '/home/tfiala/src/lldb-github/lldb/packages/Python/lldbsuite/test/lang/swift/variables/inout/a.out' (x86_64)
4321
Process 15955 stopped
* thread #1: tid = 15955, 0x000000000040140d a.out`foo(x=0x00007fffffffdfe8) -> () + 189 at main.swift:43, name = 'a.out', stop reason = breakpoint 1.1
    frame #0: 0x000000000040140d a.out`foo(x=0x00007fffffffdfe8) -> () + 189 at main.swift:43
   40
   41   func foo (inout x : Class) {
   42           print(x.ivar)
-> 43           x.ivar++ // Set breakpoint here for Class access
   44   }
   45
   46   func foo(inout x : Struct) {
(lldb) po x
[1]    15932 segmentation fault (core dumped)  ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb a.out
```

**Comment by tfiala** [comment]
Keep in mind that building LLDB on Ubuntu with the default symbol exports (i.e. only exporting public symbols from liblldb) prevents core files from containing anything useful:

```
~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb a.out -c core
(lldb) target create "a.out" --core "core"
Core file '/home/tfiala/src/lldb-github/lldb/packages/Python/lldbsuite/test/lang/swift/variables/inout/core' (x86_64) was loaded.
(lldb) bt
* thread #1: tid = 0, 0x00007fc88f6e137c, name = 'lldb', stop reason = signal SIGSEGV
  * frame #0: 0x00007fc88f6e137c
    frame #1: 0x00007fc890dce786
    frame #2: 0x00007fc89136df98
    frame #3: 0x00007fc89136e3f8
    frame #4: 0x00007fc89136e511
    frame #5: 0x00007fc891370e9d
    frame #6: 0x00007fc891370f2d
    frame #7: 0x00007fc893f38906
    frame #8: 0x00007fc8943bb718
    frame #9: 0x00007fc8943bae7a
    frame #10: 0x00007fc8943c00bd
    frame #11: 0x00007fc893f364ef
    frame #12: 0x00007fc893f366f4
    frame #13: 0x00007fc89136b048
    frame #14: 0x00007fc89136a26f
    frame #15: 0x00007fc891289f2f
    frame #16: 0x00007fc891281d2c
    frame #17: 0x00007fc890c8ba8b
    frame #18: 0x00007fc890ec594e
    frame #19: 0x00007fc890e8f63c
    frame #20: 0x00007fc890e8f86e
    frame #21: 0x00007fc890c3cac0
    frame #22: 0x00007fc890b26640
    frame #23: 0x00007fc890b26f17
    frame #24: 0x00007fc890b262ad
    frame #25: 0x00007fc890b251c6
    frame #26: 0x00007fc890c443fe
    frame #27: 0x00007fc8912d2d62
    frame #28: 0x00007fc8912d3d44
    frame #29: 0x00007fc890cb2e22
    frame #30: 0x00007fc890ca3f29
    frame #31: 0x00007fc890ca7ad6
    frame #32: 0x00007fc890ca8027
    frame #33: 0x00007fc890ba35d5
    frame #34: 0x00007fc890b78e37
    frame #35: 0x00007fc890ca8857
    frame #36: 0x00007fc88f6acde1
    frame #37: 0x0000000000409dc9 a.out`__Swift_AST + 9993
    frame #38: 0x00007ffff1ddf176
    frame #39: 0x000000000040a32f a.out`__Swift_AST + 11375
    frame #40: 0x00000000f63d4e2e
    frame #41: 0x00007fc88db5ea40
(lldb)
```
I added a mechanism back in November 2015 to allow the internal symbols of the LLDB core to be exported, which should address this.  Now I need to use that :-)

**Comment by tfiala** [comment]
> I added a mechanism back in October 2015 to allow the internal symbols of the LLDB core

And that cmake flag is '-DLLDB_EXPORT_ALL_SYMBOLS=TRUE'.  See llvm.org lldb svn trunk at r250299 for where I added it.

**Comment by tfiala** [comment]
Whoops, that was the wrong command line above:
```
$ ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb a.out -c core
```

It is lldb that crashed, so we're really trying to debug lldb, not the test inferior.  The right command line is this:
```
$ ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb -c core
```
The rebuild with -DLLDB_EXPORT_ALL_SYMBOLS=TRUE is still needed, though.

The full command line I'm using to build is:
```
$ lldb/scripts/build-swift-cmake.py --lldb-extra-cmake-args '-DLLDB_EXPORT_ALL_SYMBOLS=TRUE'
```

**Comment by tfiala** [comment]
Here's the backtrace I get at the call site:
```
$ ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb ~/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb -c core
(lldb) target create "/home/tfiala/src/lldb-github/build/Ninja-DebugAssert/lldb-linux-x86_64/bin/lldb" --core "core"
Core file '/home/tfiala/src/lldb-github/lldb/packages/Python/lldbsuite/test/lang/swift/variables/inout/core' (x86_64) was loaded.
(lldb) bt
* thread #1: tid = 0, 0x00007f7a1b0fec9c liblldb.so.3.9.0`lldb_private::Symbol::GetType(this=0x0000000000000000) const + 12 at Symbol.h:216, name = 'lldb', stop reason = signal SIGSEGV
  * frame #0: 0x00007f7a1b0fec9c liblldb.so.3.9.0`lldb_private::Symbol::GetType(this=0x0000000000000000) const + 12 at Symbol.h:216
    frame #1: 0x00007f7a1c7ec0a6 liblldb.so.3.9.0`lldb_private::Symbol::ResolveCallableAddress(this=0x0000000000000000, target=0x00000000010af3f0) const + 38 at Symbol.cpp:666
    frame #2: 0x00007f7a1cd8b8b8 liblldb.so.3.9.0`lldb_private::IRExecutionUnit::FindInSymbols(this=0x000000000268c260, specs=0x00007ffe76037c18, sc=0x000000000268c360) + 680 at IRExecutionUnit.cpp:867
    frame #3: 0x00007f7a1cd8bd18 liblldb.so.3.9.0`lldb_private::IRExecutionUnit::FindSymbol(this=0x000000000268c260, name=0x00007ffe76037c88) + 88 at IRExecutionUnit.cpp:950
    frame #4: 0x00007f7a1cd8be31 liblldb.so.3.9.0`lldb_private::IRExecutionUnit::MemoryManager::getSymbolAddress(this=0x00000000024ef220, Name="\x03v) + 81 at IRExecutionUnit.cpp:973
    frame #5: 0x00007f7a1cd8e7bd liblldb.so.3.9.0`llvm::RTDyldMemoryManager::findSymbol(this=0x00000000024ef220, Name="\x03v) + 29 at RTDyldMemoryManager.h:103
    frame #6: 0x00007f7a1cd8e84d liblldb.so.3.9.0`non-virtual thunk to llvm::RTDyldMemoryManager::findSymbol(this=0x00000000024ef220, Name="\x03v) + 29 at RTDyldMemoryManager.h:102
    frame #7: 0x00007f7a1f956226 liblldb.so.3.9.0`llvm::LinkingSymbolResolver::findSymbol(this=0x00000000026b6120, Name="\x03v) + 310 at MCJIT.cpp:643
    frame #8: 0x00007f7a1fdd9038 liblldb.so.3.9.0`llvm::RuntimeDyldImpl::resolveExternalSymbols(this=0x0000000002832810) + 504 at RuntimeDyld.cpp:821
    frame #9: 0x00007f7a1fdd879a liblldb.so.3.9.0`llvm::RuntimeDyldImpl::resolveRelocations(this=0x0000000002832810) + 506 at RuntimeDyld.cpp:93
    frame #10: 0x00007f7a1fddd9dd liblldb.so.3.9.0`llvm::RuntimeDyld::resolveRelocations(this=0x00000000026b6140) + 29 at RuntimeDyld.cpp:960
    frame #11: 0x00007f7a1f953e0f liblldb.so.3.9.0`llvm::MCJIT::finalizeLoadedModules(this=0x00000000026b5ec0) + 63 at MCJIT.cpp:229
    frame #12: 0x00007f7a1f954014 liblldb.so.3.9.0`llvm::MCJIT::finalizeObject(this=0x00000000026b5ec0) + 420 at MCJIT.cpp:253
    frame #13: 0x00007f7a1cd88968 liblldb.so.3.9.0`lldb_private::IRExecutionUnit::ReportAllocations(this=0x000000000268c260, engine=0x00000000026b5ec0) + 328 at IRExecutionUnit.cpp:1143
    frame #14: 0x00007f7a1cd87b8f liblldb.so.3.9.0`lldb_private::IRExecutionUnit::GetRunnableInfo(this=0x000000000268c260, error=0x00007ffe76038ec0, func_addr=0x0000000002190db8, func_end=0x0000000002190dc0) + 2175 at IRExecutionUnit.cpp:388
    frame #15: 0x00007f7a1cca784f liblldb.so.3.9.0`lldb_private::SwiftExpressionParser::PrepareForExecution(this=0x00000000023a2b20, func_addr=0x0000000002190db8, func_end=0x0000000002190dc0, execution_unit_sp=0x0000000002190fa0, exe_ctx=0x00007ffe76039358, can_interpret=0x0000000002190fd8, execution_policy=eExecutionPolicyOnlyWhenNeeded) + 911 at SwiftExpressionParser.cpp:1907
    frame #16: 0x00007f7a1cc9f64c liblldb.so.3.9.0`lldb_private::SwiftUserExpression::Parse(this=0x0000000002190d90, error_stream=0x00007ffe76039188, exe_ctx=0x00007ffe76039358, execution_policy=eExecutionPolicyOnlyWhenNeeded, keep_result_in_memory=true, generate_debug_info=false, line_offset=0) + 1804 at SwiftUserExpression.cpp:485
    frame #17: 0x00007f7a1c6a93ab liblldb.so.3.9.0`lldb_private::UserExpression::Evaluate(exe_ctx=0x00007ffe76039358, options=0x00007ffe760394d8, expr_cstr="// DumpForDebugger.swift\n//\n// This source file is part of the Swift.org open source project\n//\n// Copyright (c) 2014 - 2015 Apple Inc. and the Swift project authors\n// Licensed under Apache License v2.0 with Runtime Library Exception\n//\n// See http://swift.org/LICENSE.txt for license information\n// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors\n//\n// -----------------------------------------------------------------------------\n\nimport Swift\n\nfunc $__lldb__DumpForDebugger_impl<TargetStream : OutputStreamType>(\n    x_opt: Any?,\n    _ mirror: Mirror,\n    _ name: String?,\n    _ indent: Int,\n    _ maxDepth: Int,\n    _ root: Bool,\n    _ childOfCollection: Bool,\n    inout _ refsAlreadySeen : Set<ObjectIdentifier>,\n    inout _ maxItemCounter: Int,\n    inout _ targetStream: TargetStream) {\n        \n        func idForObject(x: Any) -> ObjectIdentifier? {\n            if let ao = x as? AnyObject {\n                return ObjectIdentifier(ao)\n            } else {\n                return nil\n   "..., expr_prefix=0x0000000000000000, result_valobj_sp=0x00007ffe76039558, error=0x00007ffe760392f8, line_offset=0, jit_module_sp_ptr=0x0000000000000000) + 1275 at UserExpression.cpp:263
    frame #18: 0x00007f7a1c8e326e liblldb.so.3.9.0`lldb_private::Target::EvaluateExpression(this=0x00000000010af3f0, expr_cstr="// DumpForDebugger.swift\n//\n// This source file is part of the Swift.org open source project\n//\n// Copyright (c) 2014 - 2015 Apple Inc. and the Swift project authors\n// Licensed under Apache License v2.0 with Runtime Library Exception\n//\n// See http://swift.org/LICENSE.txt for license information\n// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors\n//\n// -----------------------------------------------------------------------------\n\nimport Swift\n\nfunc $__lldb__DumpForDebugger_impl<TargetStream : OutputStreamType>(\n    x_opt: Any?,\n    _ mirror: Mirror,\n    _ name: String?,\n    _ indent: Int,\n    _ maxDepth: Int,\n    _ root: Bool,\n    _ childOfCollection: Bool,\n    inout _ refsAlreadySeen : Set<ObjectIdentifier>,\n    inout _ maxItemCounter: Int,\n    inout _ targetStream: TargetStream) {\n        \n        func idForObject(x: Any) -> ObjectIdentifier? {\n            if let ao = x as? AnyObject {\n                return ObjectIdentifier(ao)\n            } else {\n                return nil\n   "..., exe_scope=0x0000000001392a70, result_valobj_sp=0x00007ffe76039558, options=0x00007ffe760394d8) + 654 at Target.cpp:2304
    frame #19: 0x00007f7a1c8acf5c liblldb.so.3.9.0`lldb_private::SwiftLanguageRuntime::LoadDumpForDebugger(this=0x00007f7a100360b0, error=0x00007ffe76039658) + 684 at SwiftLanguageRuntime.cpp:412
    frame #20: 0x00007f7a1c8ad18e liblldb.so.3.9.0`lldb_private::SwiftLanguageRuntime::GetObjectDescription(this=0x00007f7a100360b0, str=0x00007ffe760396f8, object=0x00000000023a4520) + 62 at SwiftLanguageRuntime.cpp:461
    frame #21: 0x00007f7a1c65a3e0 liblldb.so.3.9.0`lldb_private::ValueObject::GetObjectDescription(this=0x00000000023a4520) + 480 at ValueObject.cpp:1465
    frame #22: 0x00007f7a1c543f60 liblldb.so.3.9.0`lldb_private::ValueObjectPrinter::GetDescriptionForDisplay(this=0x00007ffe76039908) + 32 at ValueObjectPrinter.cpp:186
    frame #23: 0x00007f7a1c544837 liblldb.so.3.9.0`lldb_private::ValueObjectPrinter::PrintObjectDescriptionIfNeeded(this=0x00007ffe76039908, value_printed=false, summary_printed=false) + 263 at ValueObjectPrinter.cpp:525
    frame #24: 0x00007f7a1c543bcd liblldb.so.3.9.0`lldb_private::ValueObjectPrinter::PrintChildrenIfNeeded(this=0x00007ffe76039908, value_printed=false, summary_printed=false) + 61 at ValueObjectPrinter.cpp:861
    frame #25: 0x00007f7a1c542ae6 liblldb.so.3.9.0`lldb_private::ValueObjectPrinter::PrintValueObject(this=0x00007ffe76039908) + 214 at ValueObjectPrinter.cpp:111
    frame #26: 0x00007f7a1c661d1e liblldb.so.3.9.0`lldb_private::ValueObject::Dump(this=0x000000000233eef0, s=0x00007ffe7603a5b8, options=0x00007ffe76039b90) + 62 at ValueObject.cpp:3687
    frame #27: 0x00007f7a1ccf0682 liblldb.so.3.9.0`lldb_private::CommandObjectExpression::EvaluateExpression(this=0x0000000000f44d10, expr="x", output_stream=0x00007ffe7603a5b8, error_stream=0x00007ffe7603a610, result=0x00007ffe7603a5b8) + 1170 at CommandObjectExpression.cpp:370
    frame #28: 0x00007f7a1ccf1664 liblldb.so.3.9.0`lldb_private::CommandObjectExpression::DoExecute(this=0x0000000000f44d10, command="-O  --  x", result=0x00007ffe7603a5b8) + 1908 at CommandObjectExpression.cpp:650
    frame #29: 0x00007f7a1c6d0742 liblldb.so.3.9.0`lldb_private::CommandObjectRaw::Execute(this=0x0000000000f44d10, args_string="-O  --  x", result=0x00007ffe7603a5b8) + 290 at CommandObject.cpp:1120
    frame #30: 0x00007f7a1c6c1849 liblldb.so.3.9.0`lldb_private::CommandInterpreter::HandleCommand(this=0x0000000001257ca0, command_line="po x", lazy_add_to_history=eLazyBoolCalculate, result=0x00007ffe7603a5b8, override_context=0x0000000000000000, repeat_on_empty_command=true, no_context_switching=false) + 2393 at CommandInterpreter.cpp:1834
    frame #31: 0x00007f7a1c6c53f6 liblldb.so.3.9.0`lldb_private::CommandInterpreter::IOHandlerInputComplete(this=0x0000000001257ca0, io_handler=0x00000000010ad590, line="X\xa7\x03v) + 294 at CommandInterpreter.cpp:3022
    frame #32: 0x00007f7a1c6c5947 liblldb.so.3.9.0`non-virtual thunk to lldb_private::CommandInterpreter::IOHandlerInputComplete(this=0x0000000001257ca0, io_handler=0x00000000010ad590, line="X\xa7\x03v) + 55 at CommandInterpreter.cpp:3002
    frame #33: 0x00007f7a1c5c0ef5 liblldb.so.3.9.0`lldb_private::IOHandlerEditline::Run(this=0x00000000010ad590) + 453 at IOHandler.cpp:712
    frame #34: 0x00007f7a1c596757 liblldb.so.3.9.0`lldb_private::Debugger::ExecuteIOHandlers(this=0x00000000011fcc20) + 103 at Debugger.cpp:1024
    frame #35: 0x00007f7a1c6c6177 liblldb.so.3.9.0`lldb_private::CommandInterpreter::RunCommandInterpreter(this=0x0000000001257ca0, auto_handle_events=true, spawn_thread=false, options=0x00007ffe7603a868) + 199 at CommandInterpreter.cpp:3262
    frame #36: 0x00007f7a1b0ca701 liblldb.so.3.9.0`lldb::SBDebugger::RunCommandInterpreter(this=0x00007ffe7603ac30, auto_handle_events=true, spawn_thread=false) + 129 at SBDebugger.cpp:963
    frame #37: 0x0000000000409dc9 lldb`Driver::MainLoop(this=0x00007ffe7603ac10) + 3129 at Driver.cpp:1236
    frame #38: 0x000000000040a32f lldb`main(argc=2, argv=0x00007ffe7603ae78, envp=0x00007ffe7603ae90) + 367 at Driver.cpp:1336
    frame #39: 0x00007f7a187d8a40 libc.so.6`__libc_start_main(main=(lldb`main at Driver.cpp:1304), argc=2, argv=0x00007ffe7603ae78, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007ffe7603ae68) + 240 at libc-start.c:289
    frame #40: 0x0000000000406949 lldb`_start + 41
(lldb)

```
Looking into that now.

**Comment by tfiala** [comment]
During investigation of this core dump, I did hit this:
```
(lldb) f 3
frame #3: 0x00007f7a1cd8bd18 liblldb.so.3.9.0`lldb_private::IRExecutionUnit::FindSymbol(this=0x000000000268c260, name=0x00007ffe76037c88) + 88 at IRExecutionUnit.cpp:950
   947
   948      CollectCandidateCNames(candidate_C_names, name);
   949
-> 950      lldb::addr_t ret = FindInSymbols(candidate_C_names, m_sym_ctx);
   951      if (ret == LLDB_INVALID_ADDRESS)
   952          ret = FindInRuntimes(candidate_C_names, m_sym_ctx);
   953
(lldb) p candidate_C_names
(std::vector<lldb_private::IRExecutionUnit::SearchSpec, std::allocator<lldb_private::IRExecutionUnit::SearchSpec> >) $0 = size=1 {
  [0] = {
    name = (m_string = "_TMaSp")
    mask = 2
  }
}
(lldb) p name
(const lldb_private::ConstString) $1 = (m_string = "_TMaSp")
(lldb) p m_sym_ctx
error: liblldb.so.3.9.0 DWARF DIE at 0x0a03ac67 (class ProcessInfo) has a member variable 0x0a03ac7d (m_arg0) whose type is a forward declaration, not a complete definition.
Try compiling the source file with -fno-limit-debug-info
(lldb_private::SymbolContext) $2 = {
  target_sp = {
    std::__shared_ptr<lldb_private::Target, __gnu_cxx::_Lock_policy> = {
      _M_ptr = 0x00000000010af3f0
      _M_refcount = {
...
```
Note the indication of missing debug info due to a forward declaration:
```
error: liblldb.so.3.9.0 DWARF DIE at 0x0a03ac67 (class ProcessInfo) has a member variable 0x0a03ac7d (m_arg0) whose type is a forward declaration, not a complete definition.
Try compiling the source file with -fno-limit-debug-info
```
Next time I build, I'll need to add that CFLAG.  I have enough to poke around with now, though.

**Comment by tfiala** [comment]
The candidate symbol context's symbol is NULL.  I'm not sure if that's expected, but it isn't being checked for.  I've addressed that locally and will see where that gets me.

**Comment by tfiala** [comment]
This seg fault is addressed by this change: c4cdf9657a10acc72a08bd5de5a57748e55f88ae
